File: Behavioral/Observer/CSharp/NewsAgency.cs
// NewsAgency.cs

using System;
using System.Collections.Generic;

public class NewsAgency : ISubject {
  private List<IObserver> observers = new List<IObserver>();

  public void AddObserver(IObserver observer) {
    observers.Add(observer);
  }

  public void RemoveObserver(IObserver observer) {
    observers.Remove(observer);
  }

  public void NotifyObservers(string news) {
    foreach (IObserver observer in observers) {
      observer.Update(news);
    }
  }
}

File: Behavioral/Observer/Cpp/NewsAgency.cpp
// NewsAgency.cpp

#include "NewsAgency.h"

void NewsAgency::addObserver(Observer* observer) {
  observers.push_back(observer);
}

void NewsAgency::removeObserver(Observer* observer) {
  observers.erase(std::remove(observers.begin(), observers.end(), observer), observers.end());
}

void NewsAgency::notifyObservers(const std::string& news) {
  for (Observer* observer : observers) {
    observer->update(news);
  }
}
